src/contracts/AdaptiveFeeCalculator.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: BUSL-1.1
+// This code is made available under the terms and conditions of the Business Source License 1.1 (BUSL-1.1).
+// The act of publishing this code is driven by the aim to promote transparency and facilitate its utilization for educational purposes.
+
+pragma solidity 0.8.18;
+
+import { ABDKMath64x64 } from "@hmx/libraries/ABDKMath64x64.sol";
+import { HMXLib } from "@hmx/libraries/HMXLib.sol";
+
+contract AdaptiveFeeCalculator {
+  using ABDKMath64x64 for int128;
+
+  int128 RATE_PRECISION_64x64 = ABDKMath64x64.fromUInt(1e8);
+  int128 BPS_PRECISION_64x64 = ABDKMath64x64.fromUInt(1e4);
+
+  function getAdaptiveFeeBps(
+    uint256 positionSize,
+    uint256 epochOI,
+    uint256 orderbookDepth,
+    uint256 coeffVariant,
+    uint256 baseFeeBps,
+    uint256 maxFeeBps
+  ) external view returns (uint32 feeBps) {
+    // Normalize the formula for easier coding
+    // y = 0.0007 + ((min(x/d, 1))^g) * 0.05
+    // y = 0.0007 + ((min(A, 1))^g) * 0.05
+    // y = 0.0007 + (B^g) * 0.05
+    // y = 0.0007 + C * 0.05
+    // Sell = bid
+    // Buy = ask
+
+    // x = positionSize + epochOI * 1.5
+    int128 x = _convertE8To64x64(positionSize + ((epochOI * 15) / 10));
+    int128 d = _convertE8To64x64(orderbookDepth);
+    int128 A = x.div(d);
+
+    int128 g = findG(_convertE8To64x64(coeffVariant));
+    int128 B = _min(A, ABDKMath64x64.fromUInt(1));
+    int128 C = B.pow(g);
+    int128 y = _convertE8To64x64(baseFeeBps * 1e4).add(C.mul(_convertE8To64x64(maxFeeBps * 1e4)));
+    return uint32(HMXLib.min(ABDKMath64x64.toUInt(ABDKMath64x64.mul(y, BPS_PRECISION_64x64)), uint256(maxFeeBps)));
+  }
+
+  function findG(int128 c) public pure returns (int128 g) {
+    // g = 2^(2 - min(1, c/100))
+    int128 c_100 = c.div(ABDKMath64x64.fromUInt(100));
+    int128 min = _min(ABDKMath64x64.fromUInt(1), c_100);
+    int128 expo = ABDKMath64x64.fromUInt(2).sub(min);
+    g = ABDKMath64x64.exp_2(expo);
+  }
+
+  function findC(uint256 standardDeviationE8, uint256 averagePriceE8) public view returns (int128 c) {
```

Comment:
Discussion: Should we feed `c` instead of calculate on-chain?

One value I see is we can adjust coefficient of `s` so we can control curve steepness on some markets.

Reply:
Yes, it would save computation if we do it off-chain.